### PR TITLE
EVM type refactoring

### LINKED ISSUE #1590: Remove distinction between #range and _[_.._]

### PR DESCRIPTION
**Pull request by anvacaru**
- [x] remove `_ [_ .. _]` operator (fixes #1590)
- [x] remove `Memory` sort
- [x] remove `ByteArray` sort
   - [x] remove `#sizeByteArray(_)` and use `lengthBytes(_)` instead
   - [x] remove `_ ++ _` and use `+Bytes` instead

### CODE REVIEW

File: include/kframework/lemmas/lemmas.k [hunk 1]
```diff
@@ -147,13 +146,13 @@ module LEMMAS [symbolic]
...
 
-    rule #asWord(WS) >>Int M => #asWord(WS [ 0 .. #sizeByteArray(WS) -Int (M /Int 8) ])  requires 0 <=Int M andBool M modInt 8 ==Int 0 [simplification]
+    rule #asWord(WS) >>Int M => #asWord( #range(WS, 0, #sizeByteArray(WS) -Int (M /Int 8) ))  requires 0 <=Int M andBool M modInt 8 ==Int 0 [simplification]
 
     rule #asWord(#padRightToWidth(32, BUF)) &Int notMaxUInt224 =>
          #asWord(#padRightToWidth(32, BUF))
       requires #sizeByteArray(BUF) <=Int 4 [simplification]
 
-    rule #padRightToWidth(_, BUF)[ 0 .. WIDTH ] => BUF requires #sizeByteArray(BUF) ==Int WIDTH [simplification]
+    rule #range(#padRightToWidth(_, BUF), 0, WIDTH) => BUF requires #sizeByteArray(BUF) ==Int WIDTH [simplification]
```
Comment (ehildenb) [inline include/kframework/lemmas/lemmas.k:156]: I would try just removing simplification rules like this instead of adapting them.

**Comment by ehildenb** [comment]
Some thoughts:

- Just remove `simplification` rules about `_[_.._]`, don't try to adapt them. When we see what fails we can bring back some of the lemmas needed.
- The ranged write operation `_[_:=_]` needs to be either ported to `Bytes`, or remade into something like `#writeRange(...)` (maybe @PetarMax has a preference on how to name this).
- We should be able to get rid of sorts `Memory` and `ByteArray` now, as they should just be aliases of the `Bytes` sort now (the Jvaa backend had different implementation of these sorts).

File: include/kframework/lemmas/lemmas.k (hunk 1 above)
Comment (anvacaru) [reply include/kframework/lemmas/lemmas.k]: Done.

**Comment by anvacaru** [comment]
Removed the `Memory` and `ByteArray` sort. Conformance tests are passing. I want also to remove the `++` operator since it is a duplicate of `+Bytes` currently.

File: include/kframework/evm-types.md [hunk 2]
```diff
@@ -257,10 +257,10 @@ A cons-list is used for the EVM wordstack.
...
-    rule #take(N,  BS:Bytes) => #padRightToWidth(N, .Bytes)                 requires notBool lengthBytes(BS) >Int 0 andBool         N >Int 0
-    rule #take(N,  BS:Bytes) => BS ++ #take(N -Int lengthBytes(BS), .Bytes) requires         lengthBytes(BS) >Int 0 andBool notBool N >Int lengthBytes(BS)
-    rule #take(N,  BS:Bytes) => BS [ 0 .. N ]                               requires         lengthBytes(BS) >Int 0 andBool         N >Int lengthBytes(BS)
+    rule #take(N, _BS:Bytes) => .Bytes                                          requires                                        notBool N >Int 0
+    rule #take(N,  BS:Bytes) => #padRightToWidth(N, .Bytes)                     requires notBool lengthBytes(BS) >Int 0 andBool         N >Int 0
+    rule #take(N,  BS:Bytes) => BS +Bytes #take(N -Int lengthBytes(BS), .Bytes) requires         lengthBytes(BS) >Int 0 andBool notBool N >Int lengthBytes(BS)
+    rule #take(N,  BS:Bytes) => #range(BS, 0, N)                                requires         lengthBytes(BS) >Int 0 andBool         N >Int lengthBytes(BS)
```
Comment (ehildenb) [inline include/kframework/evm-types.md:263]: I believe that this `#take` is actually no longer used. There is another `#take` for WordStack that is still used, but this one is not I believe. I'm pretty sure the same is true for `#drop` over `Bytes` type, no longer used. There is another `#drop` over wordstack that is still used, that one should stay.

File: include/kframework/evm.md [hunk 3]
```diff
@@ -2191,21 +2192,21 @@ There are several helpers for calculating gas (most of them also specified in th
...
+                 | Caddraccess    ( Schedule , Bool )                    [function, total, smtlib(gas_Caddraccess)   ]
+                 | Cstorageaccess ( Schedule , Bool )                    [function, total, smtlib(gas_Cstorageaccess)]
+                 | Csload         ( Schedule , Bool )                    [function, total, smtlib(gas_Csload)        ]
+                 | Cextcodesize   ( Schedule )                           [function, total, smtlib(gas_Cextcodesize)  ]
+                 | Cextcodecopy   ( Schedule , Int )                     [function, total, smtlib(gas_Cextcodecopy)  ]
+                 | Cextcodehash   ( Schedule )                           [function, total, smtlib(gas_Cextcodehash)  ]
+                 | Cbalance       ( Schedule )                           [function, total, smtlib(gas_Cbalance)      ]
+                 | Cmodexp        ( Schedule , Bytes , Int , Int , Int ) [function, total, smtlib(gas_Cmodexp)       ]
  // ----------------------------------------------------------------------------------------------------------------------
```
Comment (ehildenb) [inline include/kframework/evm.md:2210]: Small formatting issue.

File: include/kframework/foundry.md [hunk 4]
```diff
@@ -1076,12 +1076,12 @@ Utils
...
+    syntax KItem ::= "#setCode" Int Bytes [klabel(foundry_setCode)]
+ // ---------------------------------------------------------------
     rule <k> #setCode ACCTID CODE => . ... </k>
          <account>
            <acctID> ACCTID </acctID>
-           <code> _ => #if #asWord(CODE) ==Int 0 #then .ByteArray:AccountCode #else {CODE}:>AccountCode #fi </code>
+           <code> _ => #if #asWord(CODE) ==Int 0 #then .Bytes:AccountCode #else {CODE}:>AccountCode #fi </code>
            ...
          </account>
 ```
```
Comment (ehildenb) [inline include/kframework/foundry.md:1087]: Do we actually need this semantic kast anymore with simplified types? Can we remove the `:AccountCode` and the `{ ... }:>AccountCode` and it still builds OK?

File: kevm-pyk/src/kevm_pyk/kevm.py [hunk 5]
```diff
@@ -177,7 +172,6 @@ def concrete_rules() -> List[str]:
             'EVM-TYPES.#asByteStack',
             'EVM-TYPES.#asByteStackAux.recursive',
             'EVM-TYPES.#asWord.recursive',
-            'EVM-TYPES.ByteArray.range',
             'EVM-TYPES.bytesRange',
```
Comment (ehildenb) [inline kevm-pyk/src/kevm_pyk/kevm.py:175]: Instead of deleting, should it be updated in the same way that `concrete-rules.txt` is updated?

**Review by ehildenb — APPROVED** [review]

File: kevm-pyk/src/kevm_pyk/kevm.py (hunk 5 above)
Comment (anvacaru) [reply kevm-pyk/src/kevm_pyk/kevm.py:175]: @ehildenb This list already contains an `'EVM-TYPES.#range'`. I assumed I would duplicate it by renaming it.

File: include/kframework/foundry.md (hunk 4 above)
Comment (anvacaru) [reply include/kframework/foundry.md:1087]: It builds without an error now.

File: include/kframework/evm.md (hunk 3 above)
Comment (anvacaru) [reply include/kframework/evm.md:2210]: Done.

File: include/kframework/evm-types.md (hunk 2 above)
Comment (anvacaru) [reply include/kframework/evm-types.md:263]: Done.

File: tests/specs/concrete-rules.txt [hunk 6]
```diff
@@ -25,7 +25,7 @@ EVM.Rsstore.old
 EVM-TYPES.#asByteStack
 EVM-TYPES.#asByteStackAux.recursive
 EVM-TYPES.#asWord.recursive
-EVM-TYPES.ByteArray.range
+EVM-TYPES.Bytes.range
 EVM-TYPES.bytesRange
 EVM-TYPES.mapWriteBytes.recursive
```
Comment (ehildenb) [inline tests/specs/concrete-rules.txt:30]: Here in concrete-rules.txt we are making a change, that I don't see the same change happening in kevm.py? Are you sure we don't need to make the same change there?

File: kevm-pyk/src/kevm_pyk/kevm.py (hunk 5 above)
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/kevm.py:175]: I called out the other list I'm talking about with a comment, have a look. Basically, the concrete_rules list in `kevm.py` and the `concrete-rules.txt` file should agree with each other.

**Pull request merged**